nixos/geth: initial service

bachp opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Add a service to run a Ethereum node with go-ethereum.
The service uses the same pattern as the [bitcoind](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/networking/bitcoind.nix) and allows multiple instances to be run.

Contributes to https://github.com/NixOS/nixpkgs/issues/103890

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

bachp committed:
nixos/geth: initial service

bachp commented:
/cc @RaghavSood

bachp commented on nixos/modules/services/blockchain/ethereum/geth.nix:
```diff
@@ -0,0 +1,202 @@
+        type = types.bool;
+        default = false;
+        description = "Enable geth node.";
+      };
```
Just noticed I should have used `lib.mkEnableOption` here.

excogitation commented:
Did a quick test synching goerli. Looks to be working.
![Screenshot from 2021-01-03 01-45-46](https://user-images.githubusercontent.com/1671762/103469490-e1c8c900-4d65-11eb-8220-549ca812ad64.png)

Wouldn't it make sense to have "ipcdisable" as an option to be able to enable attaching to geth?

bachp commented:
@excogitation I ussually attacht to the node over http via `geth attach http://localhost:8545`. Adding an option would sure be possible. Do you see an advantage in having IPC too? The option could also be added later should the need arise.

As the module supports multiple instances, it might also be worth to create a wrapper script `geht-<instance>`, that has the correct parameters already set.

bachp commented:
@RaghavSood @Mic92 Could you take a look at this MR?

attila-lendvai commented on nixos/modules/services/blockchain/ethereum/geth.nix:
```diff
@@ -0,0 +1,198 @@
+      };
+
+      package = mkOption {
+        default = pkgs.go-ethereum;
```
the package is called `go-ethereum`. shouldn't we call the service the same?

bachp replied:
Looking at [other distros](https://repology.org/project/go:ethereum/versions) there seems to be no consesus if the package should be called geth or go-ethereum.

However I think we should name the service `geth` as that is the name you find in almost all guides including a popular [eth2 staking guide](https://someresat.medium.com/guide-to-staking-on-ethereum-2-0-ubuntu-lighthouse-41de20513b12). I think this is what most people expect.

I would rather propose that we rename the package to `geth` to make it more consistent.

attila-lendvai replied:
fair enough. i don't have a strong opinion on which of the two should it be, but i think being the same is preferable.

bachp replied:
The `go-ethereum` package is now renamed to `geth`.

Pinging maintainers for approval @adisbladis @lionello @xrelkd @RaghavSood

bachp replied:
Per @adisbladis comment it makes sense to keep both service and package as is. As the geth service is just one think provided by the go-ethereum package.

adisbladis commented:
I do not want to rename the package to to `geth`.
Geth is just one of the binaries in this package and the upstream repository name is `go-ethereum`.

I would be fine with an alias mapping `geth` to `go-ethereum`.

adisbladis commented on nixos/modules/services/blockchain/ethereum/geth.nix:
```diff
@@ -0,0 +1,198 @@
+      };
+
+      network = mkOption {
+        type = types.enum [ "mainnet" "goerli" "inkeby" "yolov2" "ropsten" ];
```
```suggestion
        type = types.enum [ "mainnet" "goerli" "rinkeby" "yolov2" "ropsten" ];
```

adisbladis commented on nixos/modules/services/blockchain/ethereum/geth.nix:
```diff
@@ -0,0 +1,198 @@
+      };
+
+      extraArgs = mkOption {
+        type = types.str;
```
```suggestion
        type = types.listOf types.str;
        description = "Additional arguments passed to Go Ethereum."
        default = [ ];
```

adisbladis commented on nixos/modules/services/blockchain/ethereum/geth.nix:
```diff
@@ -0,0 +1,198 @@
+        ${cfg.package}/bin/geth \
+          --nousb \
+          --ipcdisable \
+          ${optionalString (cfg.network == "goerli") "--goerli"} \
```
I'd make this one bif if statement which errors out on unknown networks, that will prevent any future programming mistakes in this module.

adisbladis replied:
Or maybe just simply `--${cfg.network}` as that seems OK for all currently defined networks.

adisbladis replied:
Or... Scratch that, this module should also be able to handle `cfg.network` being `null` which might be a reasonable thing for private networks.

bachp replied:
@adisbladis My proposal would be to change it to either: 
1. `types.nullOr types.string`
2. `types.nullOr types.enum [ "goerli" "rinkeby" "yolov2" "ropsten" ]`

Do you have a preference?

1. is more flexible as it allows any protential network without extending the service. 2. Is safer but requires to update the service in lockstep with the package.

bachp replied:
Per @adisbladis comment it makes sense to keep both service and package as is. As the geth service is just one think provided by the go-ethereum package.

adisbladis replied:
I think I prefer option 2. If you want to use a network which is not defined in the module you can always use `extraOptions`.

bachp replied:
It's now `types.nullOr (types.enum [ "goerli" "rinkeby" "yolov2" "ropsten" ])`

adisbladis commented:
@bachp I decided to make `go-ethereum` multi output in https://github.com/NixOS/nixpkgs/pull/113960 so this service doesn't need to pull in the entirety of `go-ethereum` as a runtime dependency.

bachp commented:
I did the following changes:
- package now defaults to `go-ethereum.geth`
- package is now added to the systemPackages to be able to interact with the daemon
- added test that checks if the daemon starts up with the correct network

adisbladis commented:
All good!
